Add a compiler macro to SIMPLE_EBNF to allow the ebnf syntax used in …

mingodad opened the pull request:
…https://www.bottlecaps.de/rr/ui no need of "<> <::>" wrapping production names and "// //!" for line comments and directives.

But I also want to have multiple line comments and with the existing information/documentation I'm not sure how to add it to "rpa_production_comment", can you give a help with this ?

When SIMPLE_EBNF id defined:

- <><::> is not used to wrap production names

- "//" is the single line comment

- "//!" is the way to emit directives

- "/* ... */" multi line comment is not yet implemented, need your help

Thank you for your great work !

mingodad committed:
Add a compiler macro to SIMPLE_EBNF to allow the ebnf syntax used in https://www.bottlecaps.de/rr/ui no need of <::> wrapping production names and "//" for line comments

mingodad commented:
Also the emails for contact you do not work (info_AT_rpasearch.com and bugs_AT_rpasearch.com)
Cheers !

mstoilov commented:
Hi Domingo, If you have unwrapped production names you can't have unquoted strings. Is that intended? Also I just wanted to mention that the RPA parser doesn't support regular expressions - I am not sure if you are aware of that.

Thanks,
-Martin

mingodad commented:
After converting rjs/ecma262.rap to be used in https://www.bottlecaps.de/rr/ui I've got several errors, for example on line 282 there is a missing comment:
```
11.7 Bitwise Shift Operators
```
And there is several duplicated constructs like this:
```
# 7.6 Identifier Names and Identifiers

Identifier				::= <IdentifierNameNoEmit> - (<ReservedWord> - <ReservedWord> <IdentifierPart>)
IdentifierNoEmit		::= <IdentifierNameNoEmit> - (<ReservedWord> - <ReservedWord> <IdentifierPart>)
IdentifierName			::= <IdentifierStart> <IdentifierPart>*
IdentifierNameNoEmit	::= <IdentifierStart> <IdentifierPart>*
```
What the line below mean ?
```
<IdentifierNameNoEmit> - (<ReservedWord> - <ReservedWord> <IdentifierPart>)
```

mingodad commented:
Hello Martin !
Yes it's OK to need strings always quoted.
Can you help with the "multi line" comment ?
Cheers !

mstoilov commented:
For the multi-line comment I think you need to:
1) Write rpa_production_multiline_comment() and add it in the rpa_parser_init() similar to rpa_production_comment();
2) in the rpa_production_bnf you need to add a new branch for the multi-line comment, similar to the comment branch:
	rpa_compiler_branch_begin(co, RPA_MATCH_NONE, 0);
	rpa_compiler_reference_nan_s(co, "comment");
	rvm_codegen_index_addrelocins(co->cg, RVM_RELOC_BRANCH, RPA_COMPILER_CURRENTEXP(co)->endidx, rvm_asm(RVM_BLES, DA, XX, XX, 0));
	rpa_compiler_branch_end(co);

mingodad commented:
Hello Martin !
Thank you for reply !
The problem is how to describe/write the instructions to achieve the "multi line" comments.
There is no info/docs about it, can you give a possible implementation or info/doc about it ?
Cheers !

mingodad commented:
Hello Martin !
Is this project ready for use ?
Or is it only an experiment without any actual use/application ?
Cheers !

mingodad commented:
Here is the reworked rjs/ecma262.rap that can be used on https://www.bottlecaps.de/rr/ui
[ecma262.ebnf.zip](https://github.com/mstoilov/rpatk/files/3990041/ecma262.ebnf.zip)

mstoilov commented:
I am not currently actively developing it, also I am not sure if this project is going to fit your needs, but if it does you are welcome to use it. 
Yes, the code doesn't have a lot of documentation and since I haven't touched it for years I have to figure out myself how to implement the multi line comment. It should be something like:
1. Match /*
2. Match anything that is not */
3. Match */

Similar to rpa_production_cref(), but instead of matching <: ... :>
it should match /* ... */

mstoilov commented:
here is the BNF representation of the multi line comment, you can try it at:
http://rpatk.net/web/en/parsecustom.php

`mcstart   ::= '/*'`
`mcend     ::= '*/'`
`mc        ::= <mcstart> (. - <mcend>)* <mcend>`
`<mc>`


Now, I have to remember how to implement that in the assembly language used by the RPA parser.

mingodad commented:
Hello Martin !
Thank you again for reply !
I do know the regular expression or EBNF representation of it but as you've pointed out the problem is the "assembly language used by the RPA parser".

Would be nice if in trying to do so (and remember) also added some comments to the code about it !

Also I did a mistake when I reported the problems in "rjs/ecma262.rpa" in realit the problem are in "rpagrep/ecma262.rap" they are not the same although the name is equal.
I'm looking now on the others.
Cheers !

mingodad commented:
[ecma262.ebnf.zip](https://github.com/mstoilov/rpatk/files/3990243/ecma262.ebnf.zip)
Here is the "rjs/ecma262.rap" modified to use in https://www.bottlecaps.de/rr/ui and with rpagrep compiled with "-DSIMPLE_EBNF".
When I try this:
```
./xrpagrep -f rjs/ecma262.ebnf -s "var x = 23 + (4*2);" -p
```
I'm getting this error:
```
Parse Error: Code: 2003, Name: SC at Offset: 3
```
But if remove all the directives then I get the AST, but the original "rpagrep" with "rjs/ecma262.rpa" shows nothing.
Cheers !

mingodad commented:
If we can get it to work then I plan to create a wrapper for LUA and make a small self contained executable that can be scripted to parse anything that can be described in EBNF and we can use https://www.bottlecaps.de/rr/ui or https://matthijsgroen.github.io/ebnf2railroad/try-yourself.html to document/check then.
Something like you did for PHP.

mstoilov commented:
http://rpatk.net/gitweb/?p=rpatk.git;a=commitdiff;h=1221817165ba26630de0986887631b9b951cb9a4

I added the mutiline support on a separate branch, could you please try it and let me know if it works for you.

mingodad commented:
Hello Martin !
Thank you for your help !
It seems to work when the single and multi line comments are the first declaration in a line but id doesn't when they are at the end of any other declaration:
```
var ::= "var" name ";"?   // doesn't work here
var ::= "var" name ";"?   /* doesn't work here */

```
Also could you shed a light on what this construct mean ?
```
<IdentifierNameNoEmit> - (<ReservedWord> - <ReservedWord> <IdentifierPart>)
```
The "ReservedWord" is subtracted from itself  ?

Cheers !

mingodad committed:
Add multi line comments from origin rpatk
Add a wrapper for common function calls with several identical parameters

mingodad commented:
Hello Martin !
I've added your multi line comment to my fork.
I also added some wrapper functions to group several function call with several identical parameters and I can see that there is still opportunity to add more wrappers for several common patterns and make the code more high level and easier to read and understand like :
```
	rpa_compiler_branch_begin(co, RPA_MATCH_NONE, 0);
	rpa_compiler_reference_nan_s(co, "emit");
	rvm_codegen_index_addrelocins_reloc_branch(co, RVM_BLES);
	rpa_compiler_branch_end(co);
```
There is several places where the above sequence of code appears and the only thing that changes is string parameter to "rpa_compiler_reference_nan_s", maybe something like:
```
static void rpa_compiler_branch_nan_s(const char *str)
{
	rpa_compiler_branch_begin(co, RPA_MATCH_NONE, 0);
	rpa_compiler_reference_nan_s(co, str);
	rvm_codegen_index_addrelocins_reloc_branch(co, RVM_BLES);
	rpa_compiler_branch_end(co);  
}
```
Applying the same on other places I think can improve readability and maintenance too.
Cheers !

mingodad committed:
Add more wraper functions and replace function calls with identical parameters

mingodad committed:
Add more wraper functions and replace function calls with identical parameters

mingodad committed:
Add the SIMPLE_EBNF check in another place forgotten from the initial modification
Add more wraper functions and replace function calls with identical parameters

mingodad committed:
Add a comment to explain what is implemented in a function

mingodad committed:
Add more wraper functions and replace function calls with identical parameters

mingodad committed:
Add more wraper functions and replace function calls with identical parameters

mingodad committed:
Store the return value to return it after releasing pGrep structure

mstoilov commented:
`(<ReservedWord> - <ReservedWord> <IdentifierPart>)`
means match `<ReservedWord>`, but not `<ReservedWord>` followed by `<IdentifierPart>`

`<IdentifierNameNoEmit> - (<ReservedWord> - <ReservedWord> <IdentifierPart>)`
means match `<IdentifierNameNoEmit>` , but not if matches `(<ReservedWord> - <ReservedWord> <IdentifierPart>)`

Yes, you are right adding these wrappers will definitely improve readability and maintenance. It is great you are adding them. 

I will take a look at the multi-line comment problem.

mingodad commented:
Hello Martin !
I'm reviewing my changes because the rjsexec is not working properly with my changes. I probably messed something.
Cheers !

mingodad commented:
Hello Martin !
I'm still looking to find what I messed, but I discovered that the "rjs/ecma262.rpa" as I said before has a missing comment on line 282 (see bellow) and the rjs do not seem to complain/moticed  it.
```
11.7 Bitwise Shift Operators
```
Cheers !

mstoilov commented:
Hi Domingo,

The RPA has the concept for anonymous rule, so you can do things like:
`# rpagrep -e "search for something"`

So, with the missing comment I think it just creates an anonymous production rule that is never used, but it shouldn't cause any problem.

I already mentioned before that a major drawback for RPA engine is that it doesn't support regular expressions. I added the regular expressions engine in the "rex" directory and I was planning to integrate it into the RPA, but I was never able to finish that.

mingodad commented:
Hello Martin !
And there is any command line option to deactivate "anonymous rule" to check for errors like the one mentioned above ?
Cheers !

mstoilov commented:
unfortunately no.

mingodad commented:
Hello Martin !
So the EBNF parser can not check the correctness of the grammar description ?
Cheers !

mingodad committed:
Fix incorrect wrapped functions and it's calls

mingodad commented:
Hello Martin !
I found where I messed the the refactoring in "rpa/rpaparser.c" and now rjsexec works with the small test that I was doing:
```
./rjsexec -e "var x = 23 + (3 * 45);print('hello'); print(x);"
```
But knowing that the EBNF parser do not detect grammar syntactic errors I'll think a bit before  decide if I'll go forward with this project.

Cheers !

mstoilov commented:
Hi Domingo,
You are right, the error reporting is another weak point. It is fundamentally hard to detect errors with back-tracking parser, because you if something fails to match you just don't know is it because of an error or the parser will back track and take alternative branch that will succeed.

mingodad commented:
And the debug mode does show/do something to help spot that ?

mstoilov commented:
The best way to report errors I think is to examine the parsing records and see how far the parser went before it started backtracking and eventually bail out. The stack of emitted records is still available even when the parsing fails. The records go in pair begin/end, by getting to the last 'begin' record that doesn't have a corresponding 'end' record should give you a clue what the parser attempted to match but failed.

mingodad committed:
Fixes to compile rjsexec with SIMPLE_EBNF defined

mingodad committed:
Fix rpa_dbex_create result check

mingodad commented:
Hello Martin !
Compiling rjsexec with debug info and running it through valgrind give several warnings, have you tried it ?
Cheers !

mstoilov commented:
No, I haven't used valgrind.

mingodad commented:
I'm looking at this tool http://codeworker.free.fr/index.html and it has some similarities with this project.
Did you knew it ?

mstoilov commented:
It looks interesting. No, I didn't know about it.
